quil-py/src/program/mod.rs

```diff
@@ -218,8 +229,72 @@ impl PyProgram {
             .to_owned())
     }
 
-    pub fn __str__(&self) -> String {
-        self.as_inner().to_string()
+    pub fn resolve_placeholders(&mut self) {
+        self.as_inner_mut().resolve_placeholders();
+    }
+
+    // Because we can't bubble up an error from inside the closures, they panic when the given
+    // Python functions return an error or an unexpected type. This is unusual, but in a Python
+    // program, this function will only raise because [`pyo3`] wraps Rust panics in a
+    // `PanicException`.
+    #[args("*", target_resolver = "None", qubit_resolver = "None")]
+    pub fn resolve_placeholders_with_custom_resolvers(
+        &mut self,
+        target_resolver: Option<Py<PyFunction>>,
+        qubit_resolver: Option<Py<PyFunction>>,
+    ) {
+        #[allow(clippy::type_complexity)]
+        let rs_qubit_resolver: Box<dyn Fn(&QubitPlaceholder) -> Option<u64>> =
+            if let Some(resolver) = qubit_resolver {
+                Box::new(move |placeholder: &QubitPlaceholder| -> Option<u64> {
+                    Python::with_gil(|py| {
+                        let resolved_qubit = resolver
+                            .call1(
+                                py,
+                                (placeholder
```

Comment:
This is really cool! I wonder what happens if the user-provided function calls back into rust, So Py->Rust->Py->Rust... - would there be GIL deadlock causing `Python::with_gil` to block?

Reply:
Good question! The pyQuil counterpart to this PR actually does just that and nothing locks up.